scvi/external/cell2location/_reference_module.py

```diff
@@ -0,0 +1,741 @@
+            "gene_tech_prior_beta",
+            torch.tensor(self.gene_tech_prior["alpha"] / self.gene_tech_prior["mean"]),
+        )
+
+        self.register_buffer(
+            "alpha_g_phi_hyp_prior_alpha",
+            torch.tensor(self.alpha_g_phi_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "alpha_g_phi_hyp_prior_beta",
+            torch.tensor(self.alpha_g_phi_hyp_prior["beta"]),
+        )
+        self.register_buffer(
+            "gene_add_alpha_hyp_prior_alpha",
+            torch.tensor(self.gene_add_alpha_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "gene_add_alpha_hyp_prior_beta",
+            torch.tensor(self.gene_add_alpha_hyp_prior["beta"]),
+        )
+        self.register_buffer(
+            "gene_add_mean_hyp_prior_alpha",
+            torch.tensor(self.gene_add_mean_hyp_prior["alpha"]),
+        )
+        self.register_buffer(
+            "gene_add_mean_hyp_prior_beta",
+            torch.tensor(self.gene_add_mean_hyp_prior["beta"]),
+        )
+
+        self.register_buffer("ones", torch.ones((1, 1)))
+        self.register_buffer("eps", torch.tensor(1e-8))
+
+    ############# Define the model ################
+    @staticmethod
+    def _get_fn_args_from_batch(tensor_dict):
+        x_data = tensor_dict[_CONSTANTS.X_KEY]
+        ind_x = tensor_dict["ind_x"].long().squeeze()
+        batch_index = tensor_dict[_CONSTANTS.BATCH_KEY]
+        label_index = tensor_dict[_CONSTANTS.LABELS_KEY]
+        extra_categoricals = tensor_dict[_CONSTANTS.CAT_COVS_KEY]
+        return (x_data, ind_x, batch_index, label_index, extra_categoricals), {}
+
+    def create_plates(self, x_data, idx, batch_index, label_index, extra_categoricals):
+        return pyro.plate("obs_plate", size=self.n_obs, dim=-2, subsample=idx)
+
+    def list_obs_plate_vars(self):
+        """Create a dictionary with the name of observation/minibatch plate,
+        indexes of model args to provide to encoder,
+        variable names that belong to the observation plate
+        and the number of dimensions in non-plate axis of each variable"""
+
+        return {
+            "name": "obs_plate",
+            "in": [0],  # expression data + (optional) batch index
+            "sites": {
+                "detection_y_c": 1,
+            },
+        }
+
+    def forward(self, x_data, idx, batch_index, label_index, extra_categoricals):
+
+        obs2sample = one_hot(batch_index, self.n_batch)
+        obs2label = one_hot(label_index, self.n_factors)
+        if self.n_extra_categoricals is not None:
+            obs2extra_categoricals = torch.cat(
+                [
+                    one_hot(
+                        extra_categoricals[:, i].view((extra_categoricals.shape[0], 1)),
+                        n_cat,
+                    )
+                    for i, n_cat in enumerate(self.n_extra_categoricals)
+                ],
+                dim=1,
+            )
```

Comment:
@adamgayoso is this a correct way to use categorical covariates?

Reply:
seems good to me. torch nn embedding might also be of interest, though I haven't looked carefully enough at what you're doing